### ISSUE TITLE
Add laws for the Exception class

### ISSUE THREAD

**Issue by treeowl** [haskell/core-libraries-committee#135]
Currently, the `Exception` class has no laws at all, which seems wild. I think it should have some. I believe the most obvious law is

```haskell
forall e. fromException @e . toException @e = Just
```

I think we should definitely require this. However, I was thinking about "exception sets", and those require more, so I'll explain what I mean. `Control.Exception` has a function

```haskell
catches :: IO a -> [Handler a] -> IO a
-- where
data Handler a = forall e. Exception e => Handler (e -> IO a)
```

This is totally fine when exceptions are rare, or only a few of them need to be handled. If they're common, and more than a few need to be handled, this gets slow because the whole list of handlers needs to be traversed for every exception. Due to the lawlessness of `Exception`, it is impossible to do any better without changing semantics. That seems a bit frustrating. It would be nice to be able to build a `HashMap` (or similar structure) mapping the `TypeRep` fingerprints of exception types to handlers. That way, the appropriate handler could be selected much more quickly.

The least intrusive way to support that would probably be to have a type family indicating whether `fromException` for a particular `Exception` instance `e` accepts *only* exceptions build using `SomeException @e`. This would generally be `'True`, but would be `'False` for `SomeException` and perhaps some other exotic instances.

**Comment by hasufell** [comment]
Are you proposing go change the type signature of `catches`? I'm having a hard time understanding what is the second part of your proposal.

Maybe those two parts can be separate?

**Comment by treeowl** [comment]
@hasufell Not at all. I'm suggesting it would be fun to have a package defining something approximately like

```haskell
newtype Handlers a = Handlers (DMap TypeRep (Value a))
data Value a e = Exception e => Value [e -> IO (Maybe a)]
```

**Comment by phadej** [comment]
@treeowl that doesn't work with https://hackage.haskell.org/package/base-4.17.0.0/docs/GHC-IO-Exception.html#v:asyncExceptionToException doesn't it? (as `AsyncException` is wrapped in `SomeAsyncException`) These are not exotic, `AsyncException` is one case. The exceptions thown by `postgresql-simple` will have similar subhierarchy in some future release too.

**Comment by treeowl** [comment]
@phadej I think it works fine. There's a wrapper between.

**Comment by parsonsmatt** [comment]
`fromException @e . toException @e = Just` feels pretty reasonable, and I would be surprised to see anything that fails that. The only use-case I can imagine is marking an exception as "uncatchable" - You'd have to do something very silly for that:

```haskell
data Bad = Bad
    deriving Show

instance Exception Bad where
    fromException _ = Nothing
```

This would forbid you from catching that exception specifically, but it would allow you to catch it under a `SomeException` handler. That *feels* pretty useless, though I guess I can imagine someone doing that for an asynchronous exception to ensure that it *really actually* takes down a program (along with `safe-exceptions` which avoids catching `async` exceptions in a `SomeException` match).

But, since the law doesn't forbid much, it probably doesn't buy much either. What benefit do you imagine from it?

--- 

> Not at all. I'm suggesting it would be fun to have a package defining something approximately like

You can build that package today 😄 Indeed, I've written something similar for a job queue for fun - if you squint a bit, "throwing and catching exceptions" is essentially the same thing as "requesting and handling asynchronous jobs."

I think the Trick you'll run into is with exception wrappers. `SomeException`, `SomeAsyncException`, `AnnotatedException`, `{As,S}yncExceptionWrapper`, etc will all have separate fingerprints. I've wanted to experiment with including special `fromException` behavior in some wrapper types, like:

```haskell
data MyException = FooError | BarError | OtherError SomeException

instance Exception MyException where
    fromException (SomeException inner) =
        case cast inner of 
            Just a -> Just a
            Nothing -> Just (OtherError (SomeException inner)) 

runMyThing :: IO a -> IO a
runMyThing = catch (throwIO :: MyException -> IO x)
```

which would wrap everything done in `runMyThing` in that exception type, providing a sort of "scope hierarchy."

The problem, of course, is that you then want to "nest" or "delegate" these handlers. That, to me, suggests that the proper approach is providing "unwrappers" which strip off an unwrapping layer, and then actual handlers, which can use some accumulated state from the wrappers. "Unwrappers" would just need to have a `self` reference to the final `DMap TypeRep (Value a)`, and I think that ought to tie it in alright.

The design of `Exception` is explicitly to support these wrappers and hierarchy, so anything that wants to deal with the class should take that functionality into account.

**Comment by treeowl** [comment]
@parsonsmatt, `SomeAsyncException` and `SyncExceptionWrapper` are "boring" (they use the default `fromException`). The ones on your list that do funny things are `SomeException`, `AnnotatedException`, and `AsyncExceptionWrapper`.

The simplest (not necessarily best) thing is to let each `Exception` declare via a type family whether it has a custom `fromException`. A handler set will then have to combine a map with a list. The matching order could get weird; I don't know that anything can be done about that.

**Comment by treeowl** [comment]
@Bodigrim I ask the CLC to vote on the question of whether to add the following law:

```haskell
forall e. fromException @e . toException @e = Just
```

This seems fairly uncontroversial.

**Comment by chshersh** [comment]
@treeowl Before triggering a vote, I have a single question:

* Could you check if all instances of `Exception` in `base` indeed satisfy this law?

I believe we would like to have some property-based tests for all instances. But maybe they are already trivial that we don't even need the tests to verify this.

My non-binding opinion on this proposal is +1 as I think that the law is pretty reasonable. Although, I still would like to know if it's going to be a breaking change that will change some of the instances to make them lawful.

**Comment by phadej** [comment]
> I believe we would like to have some property-based tests for all instances.

There are barely any property tests in `base`, because depending on external dependencies (e.g. `QuickCheck`) for test-suite is problematic, nor anything like that is vendored in (to be kept up with evolving GHC HEAD).
(there is `test-suite/tests/numeric/should_run/foundation.hs` which essentially vendors mini-QuickCheck into a huge single test).

There is also some doctests with `prop>`, but I don't know if they are actually run in CI. (last time I tried to run doctests locally few monhs ago they failed).  

In other words, asking for property tests is asking too much with the current state of GHC testsuite.

**Comment by parsonsmatt** [comment]
+1 from me 

The law is *really* reasonable, and anyone that wants to break it, will do so regardless of any documentation on the matter.

**Comment by mixphix** [comment]
Weakly in favour. Type class "laws" are useful sanity checks in any case.

**Comment by Bodigrim** [comment]
> Could you check if all instances of `Exception` in `base` indeed satisfy this law?

I skimmed through `base` and it seems that all of them do. The vast majority of instances just use default definitions for all methods.

----

Dear CLC members, let's vote on the proposal to add `forall e. fromException @e . toException @e = Just` as a low of `class Exception`. 
@chshersh @tomjaguarpaw @angerman @hasufell 

----

+1 from me.

**Comment by hasufell** [comment]
Isn't the law at this point just documentation? Does the CLC have to be involved in it?

**Comment by angerman** [comment]
I'm a bit lost, what actual impact does this have on anything? Isn't this primarily a documentation feature? We can define unlawful instances today, without laws being enforced?

**Comment by chshersh** [comment]
+1 from me

<hr>

The CLC documentation clearly states:

> Documentation changes normally fall under GHC developers purview, except significant ones (e. g., adding or changing type class laws)

I propose to discuss separately whether it makes sense for CLC to vote on laws docs. In my view, as I mentioned above, establishing laws may require changing implementations in a backward-incompatible way which indeed needs a CLC assessment.

Besides, documenting laws is more like establishing an expected interface and behaviour from all instances. And this may require a broader announcement to communicate changes to the Haskell users in the form of "Now you're expected to write Haskell like this".

**Comment by hasufell** [comment]
I abstain.

**Comment by tomjaguarpaw** [comment]
-1

----

I can't understand the rationale for this change.

**Comment by Bodigrim** [comment]
@angerman just a gentle reminder to vote.

**Comment by angerman** [comment]
Thanks for the ping @Bodigrim.

My concern here is around breaking changes (without notice). As I mentioned before this seems primarily informational. I'm soft +1 on this under the provision that any breaking changes resulting out of this to CLC governed libraries have to have proper deprecation cycles.

Do we have deprecation strategies for the fallout of newly defined laws?

**Comment by Bodigrim** [comment]
> Do we have deprecation strategies for the fallout of newly defined laws?

There is no immediate fallout: an introduction of a law does not give anyone a carte blanche to change existing instances bypassing normal procedures. At this point this is merely an informational change. 

----

Altogether there are 3 votes in favor, 2 softly in favor, 1 abstention and 1 vote against. This is enough for approval. @treeowl please ping me when you raise an MR.

**Comment by chshersh** [comment]
I'm trying to summarise the state of this proposal as part of my volunteering effort to track the progress of all `approved` CLC proposals.

| Field | Value |
|---|---|
| **Authors**            |  @treeowl   |
| **Status**             |  not implemented |
| **`base` version**     | Unknown  |
| **Merge Request (MR)** | none |
| **Blocked by**         | nothing |
| **CHANGELOG entry**    | missing (please, [raise an MR to `base` to update changelog][changelog])  |
| **Migration guide**    |  not needed |

[changelog]: https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/base/changelog.md

Please, let me know if you find any mistakes 🙂  

<hr>

@treeowl  could you share a status update on the implementation and what are next steps? Also, please do let CLC know if you need any help, so we can coordinate and prioritise approved proposals accordingly!
